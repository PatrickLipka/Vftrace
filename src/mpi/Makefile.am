SUBDIRS = utils

if WITH_MPI
SUBDIRS +=vftr_mpi

SUBDIRS += c2vftr \
	   c-wrapper
if ENABLE_FORTRAN
SUBDIRS += f2vftr \
	   f-wrapper
if ENABLE_FORTRAN08
SUBDIRS += f082vftr \
	   f08-wrapper
endif
endif
endif
