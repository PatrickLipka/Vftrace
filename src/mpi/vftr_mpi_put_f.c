/*
   This file is part of Vftrace.

   Vftrace is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2 of the License, or
   (at your option) any later version.

   Vftrace is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License along
   with this program; if not, write to the Free Software Foundation, Inc.,
   51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
*/

#ifdef _MPI
#include <mpi.h>

#include "vftr_mpi_put.h"

void vftr_MPI_Put_F(void *origin_addr, MPI_Fint *origin_count,
                    MPI_Fint *f_origin_datatype, MPI_Fint *target_rank,
                    MPI_Aint *target_disp, MPI_Fint *target_count,
                    MPI_Fint *f_target_datatype, MPI_Fint *f_win,
                    MPI_Fint *f_error) {

   MPI_Datatype c_origin_datatype = PMPI_Type_f2c(*f_origin_datatype);
   MPI_Datatype c_target_datatype = PMPI_Type_f2c(*f_target_datatype);
   MPI_Win c_win = PMPI_Win_f2c(*f_win);

   int c_error = vftr_MPI_Put(origin_addr,
                              (int)(*origin_count),
                              c_origin_datatype,
                              (int)(*target_rank),
                              *target_disp,
                              (int)(*target_count),
                              c_target_datatype,
                              c_win);

   *f_error = (MPI_Fint) c_error;
}

#endif
