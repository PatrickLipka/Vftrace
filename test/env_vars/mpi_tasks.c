#include <stdlib.h>
#include <stdio.h>

#ifdef _MPI
#include <mpi.h>
#endif

int main(int argc, char** argv) {
#ifdef _MPI
   MPI_Init(&argc, &argv);

   // Get number or processes
   int comm_size;
   MPI_Comm_size(MPI_COMM_WORLD, &comm_size);
   // Get rank of process
   int my_rank;
   MPI_Comm_rank(MPI_COMM_WORLD, &my_rank);

   // send/recv buffer
   int nints = 1<<my_rank;
   int *scounts = (int*) malloc(comm_size*sizeof(int));
   int *sdispls = (int*) malloc(comm_size*sizeof(int));
   int nstot = 0;
   for (int i=0; i<comm_size; i++) {
      scounts[i] = nints;
      sdispls[i] = nstot;
      nstot += scounts[i];
   }
   int *sbuffer = (int*) malloc(nstot*sizeof(int));
   for (int i=0; i<nstot; i++) {sbuffer[i]=my_rank;}

   // prepare special arrays for recv
   int *rcounts = (int*) malloc(comm_size*sizeof(int));
   int *rdispls = (int*) malloc(comm_size*sizeof(int));
   int nrtot = 0;
   for (int i=0; i<comm_size; i++) {
      rcounts[i] = 1<<i;
      rdispls[i] = nrtot;
      nrtot += rcounts[i];
   }
   int *rbuffer = (int*) malloc(nrtot*sizeof(int));
   for (int i=0; i<nrtot; i++) {rbuffer[i] = -1;}

   // Messaging
   MPI_Alltoallv(sbuffer, scounts, sdispls, MPI_INT,
                 rbuffer, rcounts, rdispls, MPI_INT,
                 MPI_COMM_WORLD);

   for (int irank=0; irank<comm_size; irank++) {
      if (my_rank == irank) {
         printf("%1d: scounts =", my_rank);
         for (int i=0; i<comm_size; i++) {
            printf(" %2d", scounts[i]);
         } printf("\n");
         printf("%1d: sdispls =", my_rank);
         for (int i=0; i<comm_size; i++) {
            printf(" %2d", sdispls[i]);
         } printf("\n");
         printf("%1d: sbuffer =", my_rank);
         for (int i=0; i<nstot; i++) {
            printf(" %2d", sbuffer[i]);
         } printf("\n");
         printf("%1d: rcounts =", my_rank);
         for (int i=0; i<comm_size; i++) {
            printf(" %2d", rcounts[i]);
         } printf("\n");
         printf("%1d: rdispls =", my_rank);
         for (int i=0; i<comm_size; i++) {
            printf(" %2d", rdispls[i]);
         } printf("\n");
         printf("%1d: rbuffer =", my_rank);
         for (int i=0; i<nrtot; i++) {
            printf(" %2d", rbuffer[i]);
         } printf("\n");
         printf("\n");
      }
      fflush(stdout);
      MPI_Barrier(MPI_COMM_WORLD);
   }

   free(sbuffer);
   sbuffer=NULL;

   free(scounts);
   scounts=NULL;

   free(sdispls);
   sdispls=NULL;

   free(rbuffer);
   rbuffer=NULL;

   free(rcounts);
   rcounts=NULL;

   free(rdispls);
   rdispls=NULL;

   MPI_Finalize();
#endif

   return 0;
}
