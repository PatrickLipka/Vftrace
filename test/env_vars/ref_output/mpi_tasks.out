0: scounts =  1  1  1  1
0: sdispls =  0  1  2  3
0: sbuffer =  0  0  0  0
0: rcounts =  1  2  4  8
0: rdispls =  0  1  3  7
0: rbuffer =  0  1  1  2  2  2  2  3  3  3  3  3  3  3  3

1: scounts =  2  2  2  2
1: sdispls =  0  2  4  6
1: sbuffer =  1  1  1  1  1  1  1  1
1: rcounts =  1  2  4  8
1: rdispls =  0  1  3  7
1: rbuffer =  0  1  1  2  2  2  2  3  3  3  3  3  3  3  3

2: scounts =  4  4  4  4
2: sdispls =  0  4  8 12
2: sbuffer =  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2
2: rcounts =  1  2  4  8
2: rdispls =  0  1  3  7
2: rbuffer =  0  1  1  2  2  2  2  3  3  3  3  3  3  3  3

3: scounts =  8  8  8  8
3: sdispls =  0  8 16 24
3: sbuffer =  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3
3: rcounts =  1  2  4  8
3: rdispls =  0  1  3  7
3: rbuffer =  0  1  1  2  2  2  2  3  3  3  3  3  3  3  3

